### ISSUE TITLE
Run a custom command before submitting a job

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#701]
In GitLab by @molidg on May 20, 2021, 16:26

Hi @dbeltrankyl @wuruchi , we have a problem to run auto-CALIOPE-urban with the operational user. The templates of auto-CALIOPE-urban are Rscripts and we need to load the module before submitting the job. To do this we included 'module load R' in the bashrc but this solution is not possible for the operational user because the operational CALIOPE crashes. It could be possible loading the module before submitting the jobs?

We had the same problem some time ago to run some python templates of auto-CALIOPE.

Thanks AS Team!

FYI: @f-macchia @gmontanepinto @charlio86 @mcastril @ jmateu

**Comment by kinow** [comment]
In GitLab by @molidg on May 20, 2021, 16:27

You can take the a3pd experiment. Let me know if you need more information

**Comment by kinow** [comment]
In GitLab by @molidg on May 20, 2021, 16:27

changed due date to May 28, 2021

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 20, 2021, 16:30

Okey, Thanks @molidg 

I'll take a look tomorrow and try to give you an answer asap

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 21, 2021, 16:42

Hello @molidg ,

I coded a vdevel with a rough version of this feature in order to see if it works.

I saw that R also fails with dummy templates(can't create COMPLETE FILE), so I tested the loading of module R using these templates on my local machine and seems that works ( although the code need some clarity work) 

I tried to test it with your experiment but I'm unable since it fails in different parts and not sure why. (also it was in my **home** local machine)

 I'm not sure what are the requisites to run this experiment, because I see that it have two different job_conf and the one that you're using in that experiment seems that have R disabled... 




What I did is to add a parameter called MODULES, defined on the jobs.conf of every (job) section that needs R.

The usage for is:
```
[job_X]

MODULES = module load R 
#MODULES = module load R ; module load other 
#MODULES = source /env.sh 
```
About marenostrum4, keep in mind that all jobs ready will be sent at once in this platform through a script, so this line will appear multiple times if there are multiple jobs, and maybe one will have to unload the previous model


**In summary, could you provide me with the steps to reproduce your experiment and the configuration for R scripts? Also is there any other module I should load besides R?**

I will test the experiment using VDEVEL autosubmit on Monday.

Thanks,

**Comment by kinow** [comment]
In GitLab by @molidg on May 21, 2021, 16:55

Hi @dbeltrankyl ! First of all thanks for your help :) I will create a small joblist to run only 1 job that needs R module loaded. We don't need more modules to run auto-CALIOPE-urban. I can help you on Tuesday (Monday is Segona Pascua :)) Thanks!!!

**Comment by kinow** [comment]
In GitLab by @f-macchia on May 21, 2021, 19:15

marked this issue as related to auto-caliope-urban#6

**Comment by kinow** [comment]
In GitLab by @molidg on May 21, 2021, 21:34

Hi @dbeltrankyl ,
I tested your solution and it's working because the R module is loaded without including 'module load R' in the bashrc. But we have the problem that for the operational user we can not use the ES modules. To do this we have to execute:

```
module use /gpfs/projects/bsc32/software/suselinux/11/modules/all
module unuse /apps/modules/modulefiles/applications /apps/modules/modulefiles/applications_bis /apps/modules/modulefiles/compilers /apps/modules/modulefiles/tools /apps/modules/modulefiles/libraries /apps/modules/modulefiles/environment /apps/modules/PRACE
```
We need to load Rv3.6.2 that is only available if we execute these lines before.

FYI: @f-macchia @gmontanepinto @charlio86 @mcastril

**Comment by kinow** [comment]
In GitLab by @molidg on May 25, 2021, 08:39

Hi @dbeltrankyl I created a short joblist for your tests. It's here:

```
/esarchive/autosubmit/a3pd/conf/jobs_a3pd_test_autosubmit.conf
```
As I said above we need to load the ES modules to use Rv3.6.2 because this version has the urbanR library that we also need.
Let me know if you need something more.

FYI: @f-macchia @gmontanepinto @charlio86 @mcastril

**Comment by kinow** [comment]
In GitLab by @gmontanepinto on May 25, 2021, 09:17

You can do a `source` of this script which has all the commands to load the modules:

`/gpfs/scratch/bsc32/bsc32530/a3j2/remoteconfig/remoteconfig.nord3.sh`

That script is sent to the scratch folder on all experiments with the SEND_SOURCE job. Would it be possible to use the Autosubmit variables to get the path of the current experiment in this new MODULES option?

**Comment by kinow** [comment]
In GitLab by @f-macchia on May 25, 2021, 10:19

We should already be able to use Autosubmit variables in the job list. I've already used them to define a custom directive in a3nd, but I can't find the documentation page where I read this feature.

**Comment by kinow** [comment]
In GitLab by @wuruchi on May 25, 2021, 11:04

Hello @f-macchia 

>We should already be able to use Autosubmit variables in the job list

It might be the feature discussed here https://earth.bsc.es/gitlab/es/autosubmit/-/issues/633

**Comment by kinow** [comment]
In GitLab by @f-macchia on May 25, 2021, 11:23

Yes!!!!

**Comment by kinow** [comment]
In GitLab by @f-macchia on May 25, 2021, 12:22

And yes, this problem could be solved using a nord3_launcher as EXECUTABLE as in https://earth.bsc.es/gitlab/es/autosubmit/-/issues/633#note_113335 (test in a3nd).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 25, 2021, 12:36

Hello @molidg ,

The actual vdevel version should work, I tested it with the job_list that you made for me, 

You can load as many things as you wish, using `;` as a separator. However don't put ANY space between the `;` there is a bug in which autosubmit is unable to get anything behind `;` if there is a space.. working on solving it before integrating it into 3.13.0b.


Example:

`MODULES = source /gpfs/scratch/bsc32/bsc32070/a3pq/remoteconfig/remoteconfig.nord3.sh;module load R`

> Would it be possible to use the Autosubmit variables to get the path of the current experiment in this new MODULES option?

Right now it is not possible, since it is directly integrated into the exec_command... It is a different approach that in the rest of the AS parameters  but will think about how to integrate %% variables. 

For now, you can use vdevel version to do test

**Comment by kinow** [comment]
In GitLab by @molidg on May 25, 2021, 14:00

ok! I will test it and let you know. Thank you very much @dbeltrankyl !

FYI: @ jmateu @mcastril @f-macchia @gmontanepinto @kserradell

**Comment by kinow** [comment]
In GitLab by @mcastril on May 26, 2021, 10:11

Good (and very efficient) job @dbeltrankyl ! I would only suggest revising the name of the `MODULES` parameter. Because I assume that you can also set other kinds of environment variables, use LD_PRELOADS, etc. Some options: PRERUN, EXPORT, ENV, PRELOAD...

**Comment by kinow** [comment]
In GitLab by @f-macchia on May 26, 2021, 11:21

mentioned in issue auto-caliope-urban#6

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2021, 17:42

Hello, 

You will be able to use Autosubmit `%var%` variables in `MODULES=` tomorrow using vdevel version. 

I expect to pass these changes to 3.13.0 ( and many others) on Monday after testing well some bugfixes and changes to other features

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on May 26, 2021, 17:44

PD: @mcastril, I still need to change the name, but unsure for which one I guess `EXPORT ` will be the chosen one

**Comment by kinow** [comment]
In GitLab by @wuruchi on May 27, 2021, 08:12

removed due date
